### ISSUE TITLE
MVP ideas

### ISSUE THREAD

**Issue by msurguy** [msurguy/svg-hatch-fill-tool#1]
Inspired by [Rough.js](https://roughjs.com/), I'd like to create a tool for SVG Hatch Fill (and other kinds of fill) of arbitrary SVG shapes.

List of features for MVP: 
- UI for uploading arbitrary SVG file
- UI for choosing fill type
- UI for adjusting fill type parameters
- A way to easily see which areas could be filled
- A way to choose each enclosed shape and apply the currently selected fill
- A way to remove the fill on any shape that had a fill applied to it
- Download the SVG with fill applied

**Comment by msurguy** [comment]
Examples of code that could be helpful for this project: 
- https://github.com/pshihn/rough/tree/master/src/fillers

**Comment by msurguy** [comment]
@pshihn was generous with his time and made the hachure-fill into separate package here: https://github.com/pshihn/hachure-fill and a demo of the package in use: https://jsbin.com/kirusaq/edit?html,output 

I'm going to figure out what kind of process we'll need to make it usable and update here.

**Comment by msurguy** [comment]
Drew process diagram of how the tool would be used:
![svg-hatch-ui](https://user-images.githubusercontent.com/585833/80786694-47124d80-8b39-11ea-8f2b-aa270a0fc280.png)

**Comment by msurguy** [comment]
Next step for me - put together UI with SVGO as the SVG optimization step (specifically for SVG shapes -> paths conversion)

**Comment by msurguy** [comment]
After thinking about this last night, this is not going to work for all use cases. Specifically, when the SVG paths in the image are not representing the areas that need to be filled. 

To address that, we need to somehow create new paths that represent areas that could be filled. 

As a proposal, we'd need to: 
- Generate a raster PNG image (maybe 2 - 3 times bigger than what's displayed on the screen) and have it under the SVG but scaled to the same dimensions as SVG
- When the user clicks within an area of the image that could be filled, determine the boundaries of the fill by raycasting or another method, creating SVG path that would be used as input for hachure-fill
- Store this path and fill details in the state to be able to quickly change fill parameters
- Keep generating paths when user interacts with the image and clicks on different areas that could be filled

**Comment by msurguy** [comment]
Here's some examples of flood fill using canvas API : 
- https://bl.ocks.org/jon-hall/2fc30039629ef22bc95c
- http://www.williammalone.com/articles/html5-canvas-javascript-paint-bucket-tool/

**Comment by msurguy** [comment]
If this is a dead end to use SVG -> Canvas -> Floodfill -> SVG path of the fill contour -> hatch fill, then maybe another approach could be used. 

What if we could determine which paths would be affected if you click inside of a fillable area, then use polygon clipper to determine the path of the area. 

Here's SVG clipper library: https://github.com/w8r/martinez
Another library is Clipper.js. More clipper libs are detailed here: https://daef.github.io/poly-bool-comparison/

**Comment by Kenno9000** [comment]
> Inspired by [Rough.js](https://roughjs.com/), I'd like to create a tool for SVG Hatch Fill (and other kinds of fill) of arbitrary SVG shapes.

Hi, 
It looks similar to Lionel's solution, which I can say, plots very well.  
Code/product available on  obseravble
Lionel Radisson  makio135.com
https://observablehq.com/@makio135/hatch-test-cmj-n
see:
Hatch test CMJ(N)
Möbius strips for pen plotter

Thanks for all your efforts, I have been able to move ahead in this area.  It wouldn't be the same without you.

**Comment by msurguy** [comment]
Thanks @Kenno9000 ! Yes, there are ways to generate the hatching between the paths. Because the goal is to make a tool where the user would be able to click any point within SVG and it would fill the area like a paint bucket tool would in Photoshop or MSPaint, the difficult part is to generate those boundaries for the fill. Basically determining that contour is the hard part

**Comment by cleversomeday** [comment]
feature request ... the ability to invert the fill for drawing white on black or for engraving.
random data point... the paint bucket tool in Inkscape behaves exactly as you describe to create a closed vector path inside a clicked area.

**Comment by msurguy** [comment]
@cleversomeday thanks for the info and the feature request! Indeed, the Inkscape bucket fill looks to be operating almost exactly the way I thought it should, so now I know I'm on the right track...

From Inkscape docs about the bucket tool:
<img width="1166" alt="Screen Shot 2020-05-05 at 2 54 43 PM" src="https://user-images.githubusercontent.com/585833/81136965-e6f12200-8f11-11ea-877e-9023cae4b3ac.png">

I wonder how the inverse of the fill might work, since my eventual goal is to be able to go beyond hatch and line patterns. I'll think about it though.

**Comment by msurguy** [comment]
@Kenno9000 indeed, Lionel has contributed to me figuring this out! Using a couple of his tools, I was able to make what I wanted! 

Here's the tool so far: https://observablehq.com/@msurguy/vector-fill

**Comment by msurguy** [comment]
@cleversomeday feel free to fork the Observable Notebook, replace the SVG code and see if it's working for you! https://observablehq.com/@msurguy/vector-fill 

I'm going to start on the UI for this tool soon.
